core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -413,6 +414,34 @@ public void writeMap(@Nullable Map<String, Object> map) throws IOException {
     }
 
     /**
+     * write map to stream with consistent order
+     * to make sure every map generated bytes order are same
+     */
+    public void writeMapWithConsistentOrder(@Nullable Map<String, Object> map) throws IOException {
```

Comment:
I like how this makes reading the read and write side easy. Can you add a javadoc to the method saying it is compatible with `readMap` and `readGenericValue`?

Can you add a dedicated test case for this method in `BytesStreamsTests`?

I think we shouldn't try to handle sorting recursively in this, at least for now. Can you say something about only sorting the keys of the map, not maps contained within the map?

Maybe the signature for the method should be `public <K extends Comparable<K>> void writeMapWithConsistentOrder(@Nullable Map<K, ? extends Object> map) throws IOException`? That way it can function on `Map<String, String>` and `Map<Integer, String>` or any other weirdness we need to throw at it.